### ISSUE TITLE
how to reconstructe by multiple GPU

### ISSUE THREAD

**Issue by ccs1019** [chiahao3/ptyrad#3]
I wonder how to use multiple GPU devices to accelerate the calculation?

**Comment by chiahao3** [comment]
Hi, thanks for the interest in **PtyRAD**!

We're still working on the `docs/` but I can give a brief instruction here.

I assume you meant "using multiple GPUs for 1 reconstruction task", because PtyRAD can also handle multiple GPU workers for multiple trials during hyperparameter tuning (hypertune) mode. For more information of the hypertune mode, you might want to take a look of my recent tutorial video [here](https://cornell.box.com/s/n5balzf88jixescp9l15ojx7di4xn1uo). This Box link would be added to README in our next commit, but I'll share it here as a reference. Note that the previous summary slides could be a bit outdated.

To reconstruct with multiple GPUs, you'll need the following:

1. Linux system
2. more than 1 NVIDIA GPU on the same machine (multi-instance GPU, or MIGs are NOT considered as multi GPU and would not work, see [here](https://discuss.pytorch.org/t/parallel-training-with-invidia-migs/159445).)
3. `accelerate` package from HuggingFace (it's currently listed as a core dependency of PtyRAD so if you follow the README instruction while creating environment, it should already be included)

If you meet all the (mostly hardware) requirement, then you can simplify run these in your terminal:

`PARAMS_PATH="params/tBL_WSe2_reconstruct.yml"`
`accelerate launch --multi_gpu --num_processes=2 --mixed_precision='no' ptyrad run --params_path "${PARAMS_PATH}"`

See this demo [Slurm script](https://github.com/chiahao3/ptyrad/blob/main/scripts/slurm_run_ptyrad.sub) for more information.

It's important to note that **multi GPU doesn't always speed up the reconstruction** because it splits the mini-batch to multiple GPU devices so there're inter-communication overhead and generally it's only beneficial to do multi GPU when your batch size (or detector pixel size) is very big as discussed in our [arXiv paper](https://arxiv.org/abs/2505.07814). For the tested gradient descent optimizers like Adam, usually smaller batch size leads to better final quality. Overall, despite **PtyRAD** supports multi GPU capability now, we expect it'll become more useful in the future with larger detector pixels or for ptycho-tomo tasks.

![Image](https://github.com/user-attachments/assets/88224fe8-0820-490e-8472-abbae166d58c)

**Comment by ccs1019** [comment]
---

Thank you for your prompt response and generous sharing of information. This is highly valuable for advancing the electron ptychography community! Ptyrad is truly groundbreaking. So far, I have noticed some differences between Ptyrad and Foldslice, as certain parameters seem to have non-equivalent effects. I hope to reach a clear conclusion to better clarify their differences for further comparison.

---
